python/ql/src/experimental/dataflow/internal/DataFlowUtil.qll

```diff
@@ -45,8 +45,14 @@ EssaNode importModule(string name) {
       or
       name = alias.getValue().(ImportExpr).getImportedModuleName()
     ) and
-    result.getVar().(AssignmentDefinition).getSourceVariable() = var
+    result.(EssaNode).getVar().(AssignmentDefinition).getSourceVariable() = var
   )
+  or
+  // In `from module import attr`, we want to consider `module` to be an expression that refers to a
+  // module of that name, as this allows us to refer to attributes of this module, even if it's
+  // never imported directly. Note that there crucially isn't any _flow_ from `module` to references
+  // to that same identifier.
```

Comment:
```suggestion
  // In `from module import attr`, we want to consider `module` to be an expression that refers to a
  // module of that name, as this allows us to model the whole import as an `AttrRead` (although
  // the module is not imported directly). Note that there crucially isn't any _flow_ from `module`
  // to references to that same identifier.
```

Reply:
I don't think this rewrite is an improvement, as it makes it sound as if we wouldn't be able to model imports as `AttrRead`s otherwise (which is false -- the problem lies with type tracking).

Reply:
Ahh yes, it does sound rather confusing. I just wanted us to highlight the context of _why_ this extra thing was important.

To get this import working properly with `AttrRead` and type-tracking, we need to expose `module` 

I'm lacking the use of keywords `AttrRead` and "type-tracking" in the current comment. I guess what I'm aiming for is something like the following:

> To get this `AttrRead` working properly with type-tracking, we need to expose the "object" of the `AttrRead` as an import of that module.

Reply:
Honestly, I still don't think that's needed. In fact, I think the comment itself is superfluous (but I added it in anticipation of some confusion).

What I mean is: the predicate states that it holds for `Node`s that hold a reference to a module with the given name. (An `EssaNode`, in fact, and this _does_ need fixing.) The disjunct I added does not change this in any way -- `from foo import bar` _does_ indeed refer to a module named `foo`. Rather, the previous implementation was incomplete, and now it's more complete.

Reply:
> the predicate states that it holds for Nodes that hold a reference to a module with the given name.

That's not what it says right now, but it probably should :smile:

> The disjunct I added does not change this in any way -- `from foo import bar` _does_ indeed refer to a module named `foo`. Rather, the previous implementation was incomplete, and now it's more complete.

Let's update the [QLDoc](https://github.com/github/codeql/pull/4423/files/df447c0af9ec7d62b428ebdfb5d21efcad2c1203..d46453caaa6dd28ab0ed5a183fa077e4e17f4c61#diff-d51d1db0971149bad11ef4f27a9cb3beR25-R28) that explains what is handled by this predicate then :wink: 

> Honestly, I still don't think that's needed.

I absolutely do! I think it requires non-trivial insight into other parts of our library to realize why this disjunct is required, and I would very much like a comment highlighting this.